### ISSUE TITLE
KotlinxSerializer unable to parse lists

### ISSUE THREAD

**Issue by viteinfinite** [ktorio/ktor#906]
### Ktor Version
1.1.2

### Ktor Engine Used(client or server and name)
iOS Client

### JVM Version, Operating System and Relevant Context
iOS 12

### Feedback
Hey, the KotlinxSerializer is apparently unable to deserialized JSON payloads having lists at the root.
For instance, a JSON payload like the following:
```json
[{"foo": "bar1"}, {"foo": "bar2}]
```
would not be deserialized correctly.

After some trial and error, the problem lies on `KotlinSerializer.kt:94`:
```kotlin
    private fun lookupSerializerByType(type: KClass<*>): KSerializer<*> {
        mappers[type]?.let { return it } // <--
        return (type.defaultSerializer() ?: type.serializer())
    }
```

From what I understand, no mapper will be found because `type` would be `kotlin.collections.List`. Is that correct?

Thank you in advance :)

**Comment by e5l** [comment]
Hi @viteinfinite, thanks for the report.

Did you try `registerList` to register top-level list serializers?

**Comment by philmitchell** [comment]
`registerList()` is not working for me. **ktor** deserializes single `Todo` fine, but for list throws: `Can't locate argument-less serializer for class kotlin.collections.List. For generic classes, such as lists, please provide serializer explicitly.`

```kotlin
    val httpClient = HttpClient() {
        install(JsonFeature) {
            serializer = KotlinxSerializer().apply {
                registerList(Todo.serializer().list)
            }
        }
    }

    suspend fun getTodo(): Todo = coroutineScope {
        val urlString = "https://jsonplaceholder.typicode.com/todos/1"
        httpClient.get<Todo> {
            url(urlString)
        }
    }

    suspend fun getTodos(): List<Todo> = coroutineScope {
        val urlString = "https://jsonplaceholder.typicode.com/todos"
        httpClient.get<List<Todo>> {
            url(urlString)
        }
    }
```

**Comment by e5l** [comment]
Could you try `registerList(Todo.serializer())` instead?

**Comment by philmitchell** [comment]
Thanks, tried it but still got same result.

**Comment by hikkidev** [comment]
Are there any advancements? (

**Comment by e5l** [comment]
The fix should be available from kotlin `1.3.50`

**Comment by bernatpl** [comment]
Parsing lists is already working in ktor version 1.3.0 (currently beta). In this new version there is no need to register the serializers.

**Comment by e5l** [comment]
Fixed with `typeOf` support in 1.3.0
